### DISCUSSION TITLE
JIT mode breaks build in NextJS

### DISCUSSION THREAD

**Discussion by HumidBrains** [tailwindlabs/tailwindcss#5572] [Help] [answered]
**What version of Tailwind CSS are you using?**
Tailwindcss v2.2.15

**What build tool (or framework if it abstracts the build tool) are you using?**
NextJS v11.1.2

**What version of Node.js are you using?**
v16.6.1

**What operating system are you using?**
macOS Big Sur v11.5.2 

**Describe your issue**

Wanted to try JIT mode in my functioning, jet very slow Tailwind setup.
Set mode to 'jit' in tailwind.config.js and ran yarn dev. This results in the error below, so not specifically a font related bug if it's a bug. If not a bug, what may I have done wrong?

Sorry about the messy code blocks below, not sure why the markdown won't wrap the code properly. 
Grateful for any help.

```
event - build page: /next/dist/pages/_error
error - ./node_modules/next/dist/compiled/css-loader/cjs.js??ruleSet[1].rules[3].oneOf[5].use[1]!./node_modules/next/dist/compiled/postcss-loader/cjs.js??ruleSet[1].rules[3].oneOf[5].use[2]!./node_modules/@fontsource/inter/index.css
TypeError: plugin is not a function
    at runMicrotasks (<anonymous>)
```

As the error is related to an imported font, I removed it only to get the next error.

```
error - ./node_modules/next/dist/compiled/css-loader/cjs.js??ruleSet[1].rules[3].oneOf[7].use[1]!./node_modules/next/dist/compiled/postcss-loader/cjs.js??ruleSet[1].rules[3].oneOf[7].use[2]!./node_modules/next/dist/compiled/resolve-url-loader/index.js??ruleSet[1].rules[3].oneOf[7].use[3]!./node_modules/next/dist/compiled/sass-loader/cjs.js??ruleSet[1].rules[3].oneOf[7].use[4]!./styles/main.scss
```

**My Tailwind config**

```js
const tailwindcss = require('tailwindcss')
const precss = require('precss')
const postcss = require('postcss')
const forms = require('@tailwindcss/forms')
const typography = require('@tailwindcss/typography')
const autoprefixer = require('autoprefixer')
const colors = require('tailwindcss/colors')
const lineClamp = require('@tailwindcss/line-clamp')
const aspectRatio = require('@tailwindcss/aspect-ratio')

module.exports = {
  mode: 'jit',
  purge: ['./components/**/*.js', './pages/**/*.js'],
  theme: {
    extend: {
      fontFamily: {
        sans: `Inter, -apple-system, BlinkMacSystemFont`,
      },
    },
    variants: {
      extend: {
        lineClamp: ['hover'],
      },
    },
  },
  variants: {},
  plugins: [
    tailwindcss,
    postcss,
    autoprefixer,
    precss,
    forms,
    typography,
    colors,
    lineClamp,
    aspectRatio,
  ],
}
```

**My NextJS config**

```js
module.exports = {
  typescript: {
    ignoreBuildErrors: true,
  },
  eslint: {
    ignoreDuringBuilds: true,
  },
  trailingSlash: true,
  webpackDevMiddleware: (config) => {
    config.watchOptions = {
      poll: 1000,
      aggregateTimeout: 300,
    }

    return config
  },
  sassOptions: {
    includePaths: [path.join(__dirname, 'styles')],
  },
  images: {
    domains: [
      allowedImageWordPressDomain,
      'via.placeholder.com',
      'placeimg.com',
    ],
  },
}
```

**Comment by reinink** [comment, marked as answer]
Hey @HumidBrains, happy to help out here.

In these situations we really need a *minimum* reproduction repo, otherwise it's very hard for us to reproduce the issue ourselves. Can you please provide that?

One thing I can see that's wrong in your `tailwind.config.js` file is the `plugins` property. You have a bunch of things listed here that are not valid Tailwind CSS plugins.

```js
plugins: [
  tailwindcss, // not a valid plugin
  postcss, // not a valid plugin
  autoprefixer, // not a valid plugin
  precss, // not a valid plugin
  forms,
  typography,
  colors,  // not a valid plugin
  lineClamp,
  aspectRatio,
],
```

For the colors, you can see how to enable the extra colors [here](https://tailwindcss.com/docs/customizing-colors).

The rest (`tailwindcss`, `postcss`, `autoprefixer` and `precss`) are all meant to be run outside of this file — part of your PostCSS build step. I'd recommend having a look at this [example app](https://github.com/vercel/next.js/tree/canary/examples/with-tailwindcss) to learn how to setup Tailwind CSS in Next.js. You might also find [these instructions](https://nextjs.org/docs/advanced-features/customizing-postcss-config) from the Next.js docs helpful.

**Reply by HumidBrains** [reply]
Hey, @reinink thanks for your kind fast reply and sorry for my slow response. The colors work fine, but I had removed stuff from the config to shorten it a bit. I had a go with the setup in the NextJS example app and it worked out fine. However, I got too many layout issues migrating to JIT due to some classes no longer obeying me, so have to keep on suffering compile times for now. ;)
